src/aviary/functional.py

```diff
@@ -0,0 +1,175 @@
+from collections.abc import Callable
+from functools import update_wrapper
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from aviary.env import Environment, Frame
+from aviary.message import Message
+from aviary.tools import Messages, Tool, ToolRequestMessage
+from aviary.utils import is_coroutine_callable
+
+
+class DynamicState(BaseModel):
+    """Dynamic env state model that adapts to provided extras."""
+
+    reward: float = 0
+    done: bool = False
+    extras: dict[str, Any] = Field(default_factory=dict)
+
+    def __getattr__(self, name: str):
+        """Allow direct access to extras as attributes."""
+        if name in self.extras:
+            return self.extras[name]
+        raise AttributeError(f"'State' has no attribute {name!r}")
+
+
+class FunctionalEnvironment(Environment[DynamicState]):
+    """Environment class for function-based environments."""
+
+    def __init__(
+        self,
+        start_fn: Callable,
```

Comment:
```suggestion
        start_fn: Callable[..., tuple[str, dict[str, Any]]],
```

Will help people hit the `TypeError` blow up in `_call_start_fn` less

Reply:
This doesn't seem to work because of the possibility of it being awaitable. Couldn't easily find the correct type - the one you suggested fails.